### PR TITLE
[debugger] Clean up breakpoint state for dead jobs

### LINKED ISSUE #17040: [Ray debugger] Clean up debugging state after job finishes
### What is the problem?

Currently the active breakpoints in the debugger are not cleaned up after the job that set the breakpoints exists. We should clean them up.

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by edoakes**
## Related issue number

Closes https://github.com/ray-project/ray/issues/17040

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/scripts/scripts.py [hunk 1]
```diff
@@ -217,7 +218,10 @@ def debug(address):
     logger.info(f"Connecting to Ray instance at {address}.")
     ray.init(address=address, log_to_driver=False)
     while True:
-        continue_debug_session()
+        # Used to filter out and clean up entries from dead jobs.
+        live_jobs = set(
+            j["JobID"] for j in ray.state.jobs() if not j["IsDead"])
```
Comment (pcmoritz) [inline python/ray/scripts/scripts.py:223]: consider naming `j` -> `job` (`j` sounds more like a loop index)

**Review by pcmoritz — APPROVED** [review]
thx for fixing this!

**Comment by pcmoritz** [comment]
Ah one thing I noticed while running throught this example here: https://docs.ray.io/en/master/ray-debugging.html#stepping-between-ray-tasks

It gave me an error in line 187 of scripts.py (TypeError: byte indeices must be integers or slices, not str)

**Comment by pcmoritz** [comment]
I think the "exit_debugger" entry is handled differently from the rest and is just a string instead of a dict. I think just removing the `live_jobs` handling from `continue_debug_session` makes sense here (it means that if a job terminates while you are debugging it you might get an error, but that's already the case because the debugger you are currently attached to will go away).

**Comment by edoakes** [comment]
@pcmoritz ah, good catch, thanks for testing it out. Your proposal makes sense, I guess in this case you might see an error once but then the entry should be deleted by the existing logic [here](https://github.com/ray-project/ray/blob/master/python/ray/scripts/scripts.py#L191), right?

Actually, do we need to reorder lines 190 and 191? Otherwise I think the connect_pdb_client call will fail and the KV entry won't actually be deleted.

**Comment by edoakes** [comment]
Looks like this is hanging while trying to "continue" on a dead job's breakpoint:


```
(base) C02X23BRJHD3:client eoakes$ ray debug
/Users/eoakes/code/ray/python/ray/autoscaler/_private/cli_logger.py:57: FutureWarning: Not all Ray CLI dependencies were found. In Ray 1.4+, the Ray CLI, autoscaler, and dashboard will only be usable via
`pip install 'ray[default]'`. Please update your install command.
  warnings.warn(
2021-07-15 13:56:03,892 INFO scripts.py:217 -- Connecting to Ray instance at 192.168.0.132:6379.
2021-07-15 13:56:03,893 INFO worker.py:739 -- Connecting to existing Ray cluster at address: 192.168.0.132:6379
Continuing pdb session in different process...
... hanging ...
... hanging ...
... hanging ...
... hanging ...
... hanging ...
```

**Comment by edoakes** [comment]
Ok @pcmoritz I updated the logic to also include the job_id in the RAY_PDB_CONTINUE entries to address the above, PTAL.

**Comment by pcmoritz** [comment]
Thanks Edward, looks good to me, I'm trying it out now and then we can merge it :)

**Comment by pcmoritz** [comment]
The recursive debugging is working for me now, looks good to merge!

**Comment by edoakes** [comment]
Failures are java and `test_kill_raylet_signal_log`, which both look unrelated.

**Review by simon-mo — APPROVED** [review]

**Pull request merged**